File: app/src/main/java/com/github/alexeygorovoy/moxy_template/utils/rx/RxSchedulers.java
package com.github.alexeygorovoy.moxy_template.utils.rx;

import rx.Completable;
import rx.Observable;
import rx.Scheduler;
import rx.Single;


public abstract class RxSchedulers {


    abstract public Scheduler mainThread();

    abstract public Scheduler io();

    abstract public Scheduler computation();

    public <T> Observable.Transformer<T, T> ioToMain() {
        return objectObservable -> objectObservable
                .subscribeOn(io())
                .observeOn(mainThread());
    }

    public Completable.Transformer ioToMainCompletable() {
        return completable -> completable
                .subscribeOn(io())
                .observeOn(mainThread());
    }

    public <T> Single.Transformer<T, T> ioToMainSingle() {
        return objectObservable -> objectObservable
                .subscribeOn(io())
                .observeOn(mainThread());
    }

    public <T> Observable.Transformer<T, T> computationToMain() {
        return objectObservable -> objectObservable
                .subscribeOn(computation())
                .observeOn(mainThread());
    }

    public <T> Single.Transformer<T, T> computationToMainSingle() {
        return objectObservable -> objectObservable
                .subscribeOn(computation())
                .observeOn(mainThread());
    }
}

File: app/src/main/java/com/github/alexeygorovoy/moxytemplate/rx/RxSchedulers.kt
package com.github.alexeygorovoy.moxytemplate.rx

import io.reactivex.*


abstract class RxSchedulers {

    abstract fun mainThread(): Scheduler

    abstract fun io(): Scheduler

    abstract fun computation(): Scheduler

    fun <T> ioToMain(): ObservableTransformer<T, T> {
        return ObservableTransformer { objectObservable ->
            objectObservable
                .subscribeOn(io())
                .observeOn(mainThread())
        }
    }

    fun ioToMainCompletable(): CompletableTransformer {
        return CompletableTransformer { completable ->
            completable
                .subscribeOn(io())
                .observeOn(mainThread())
        }
    }

    fun <T> ioToMainSingle(): SingleTransformer<T, T> {
        return SingleTransformer { objectObservable ->
            objectObservable
                .subscribeOn(io())
                .observeOn(mainThread())
        }
    }

    fun <T> computationToMain(): ObservableTransformer<T, T> {
        return ObservableTransformer { objectObservable ->
            objectObservable
                .subscribeOn(computation())
                .observeOn(mainThread())
        }
    }

    fun <T> computationToMainSingle(): SingleTransformer<T, T> {
        return SingleTransformer { objectObservable ->
            objectObservable
                .subscribeOn(computation())
                .observeOn(mainThread())
        }
    }
}
